packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,119 @@
+import { IResource } from './resource';
+import { Token } from "./token";
+
+/**
+ * Represents the human-readable, physical name of a resource.
+ * These can be known at code runtime,
+ * in the case of fixed, customer-provided names,
+ * at synthesis time, in case of automatically assigned names by the framework,
+ * or only at deploy time, assigned by CloudFormation -
+ * which is the default.
+ *
+ * @see #deployTime()
+ * @see #deployTimeOrAssigned()
+ * @see #of()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically assigned at deploy time.
+   * This is the default if you don't specify a physical name.
+   */
+  public static deployTime(): PhysicalName {
+    return new DeployTimePhysicalName();
+  }
+
+  /**
+   * A physical name that will be assigned at synthesis time if the resource
+   * is used in a cross-environment manner,
+   * and at deploy time otherwise.
+   * This is the default name implicitly created sub-resources
+   * (like IAM Roles) are created with in the Construct Library.
+   * If you're using a Construct in a cross-environment manner,
+   * you need to use either this name,
+   * or a {@link #of() fixed name}.
+   */
+  public static deployTimeOrAssigned(): PhysicalName {
+    return new DeployTimeOrAssignedPhysicalName();
+  }
+
+  /**
+   * A fixed physical name (one that is known statically, at synthesis time).
+   *
+   * @param name the name to assign
+   */
+  public static of(name: string): PhysicalName {
+    return new FixedPhysicalName(name);
+  }
+
+  protected constructor() {
+  }
+
+  /**
+   * Returns the physical name, as a String.
+   * Can be `undefined`, if a physical name has not been provided,
+   * or an opaque Token, if a generated name should be assigned lazily at synthesis time.
+   */
+  public abstract asString(): string | undefined;
+
+  /** @internal */
+  public abstract _resolveForCrossEnv(resource: IResource): void;
+}
+
+// implementations are private to this module,
+// we only surface them through static factory methods
+
+class DeployTimePhysicalName extends PhysicalName {
+  public asString(): string | undefined {
+    return undefined;
+  }
+
+  /** @internal */
+  public _resolveForCrossEnv(resource: IResource): void {
+    // error out - a deploy-time name cannot be used across environments
+    throw new Error(`Cannot use resource '${resource.node.path}' in a cross-environment fashion, ` +
```

Comment:
Provide instructions on how to fix this

Reply:
Can you propose some wording? If we had a property `physicalName` for all Resources, I could see mentioning it here, but since you prefer `bucketName`, `roleName`, etc., I'm not sure what should the wording be.

Reply:
Done.